test/WakuRlnV2.t.sol

```diff
@@ -789,4 +881,419 @@ contract WakuRlnV2Test is Test {
+
+        // Erase and check event
+        vm.expectEmit(true, true, true, true);
+        emit MembershipUpgradeable.MembershipExpired(idCommitment, rateLimit, 0);
+        wZeroGrace.eraseMemberships(ids);
+
+        (,,,, uint32 fetchedRateLimit,,,) = wZeroGrace.memberships(idCommitment);
+        assertEq(fetchedRateLimit, 0);
+    }
+
+    function test__FullCleanUpErasure() external {
+        uint256 idCommitment = 1;
+        uint32 rateLimit = w.minMembershipRateLimit();
+        (, uint256 price) = w.priceCalculator().calculate(rateLimit);
+
+        token.approve(address(w), price);
+        w.register(idCommitment, rateLimit, noIdCommitmentsToErase);
+
+        uint256 initialRoot = w.root();
+
+        (
+            , // depositAmount
+            , // activeDuration
+            uint256 graceStart,
+            uint32 gracePeriodDuration,
+            , // rateLimit
+            , // index
+            , // holder
+                // token
+        ) = w.memberships(idCommitment);
+
+        vm.warp(graceStart + gracePeriodDuration + 1); // Expire
+
+        uint256[] memory toErase = new uint256[](1);
+        toErase[0] = idCommitment;
+        w.eraseMemberships(toErase, true); // Full clean-up
+
+        // Use public function to get rate commitment at index 0
+        uint256[] memory commitments = w.getRateCommitmentsInRangeBoundsInclusive(0, 0);
+        assertEq(commitments[0], 0);
+
+        assertNotEq(w.root(), initialRoot); // Root changed
+
+        // Count the length of indicesOfLazilyErasedMemberships
+        uint256 erasedLength = 0;
+        while (true) {
+            try w.indicesOfLazilyErasedMemberships(erasedLength) {
+                erasedLength++;
+            } catch {
+                break;
+            }
+        }
+        assertEq(erasedLength, 1);
+        assertEq(w.nextFreeIndex(), 1); // Unchanged
+    }
+
+    function test__TokenTransferFailures() external {
+        // Deploy MaliciousToken implementation
+        MaliciousToken maliciousTokenImpl = new MaliciousToken();
+
+        // Deploy proxy with no reentrancy (enables failTransfer)
+        ERC1967Proxy proxy = new ERC1967Proxy(
+            address(maliciousTokenImpl), abi.encodeCall(MaliciousToken.initialize, (address(0), "", true))
+        );
+        MaliciousToken maliciousToken = MaliciousToken(address(proxy));
+
+        // Mint tokens
+        maliciousToken.mint(address(this), 100_000_000 ether);
+
+        // Set price calculator
+        vm.prank(w.owner());
+        w.setPriceCalculator(address(new DeployPriceCalculator().deploy(address(maliciousToken))));
+
+        uint32 rateLimit = w.minMembershipRateLimit();
+        (, uint256 price) = w.priceCalculator().calculate(rateLimit);
+
+        // Approve tokens
+        maliciousToken.approve(address(w), price);
+
+        // Expect transfer failure
+        vm.expectRevert("Malicious transfer failure");
+        w.register(1, rateLimit, noIdCommitmentsToErase);
+    }
```

Comment:
While this test checks if malicious token does indeed revert, it's not actually testing any functionality on the rln contract itself. 

So I'm not sure it adds any value.

Reply:
It is from similar category like were the reentrancy tests. If during deployment or reconfiguration owner enters by accident wrong address with malicious token contract, it could lead to lost funds. There is a small chance, when the deployment scripts sit on public repo, that repo will get hacked and deployment will happen to malicious contract.

Reply:
I've cleanup up the MaliciousToken from reentrancy capabilities. https://github.com/waku-org/waku-rlnv2-contract/pull/31/commits/fb9f55b6cd63de22c1702ffba067019ed49f911a